vowpalwabbit/core/src/reductions/shared_feature_merger.cc

```diff
@@ -64,12 +67,18 @@ void predict_or_learn(sfm_data& data, VW::LEARNER::learner& base, VW::multi_ex&
 
   // Guard example state restore against throws
   auto restore_guard = VW::scope_exit(
-      [has_example_header, &shared_example, &ec_seq, &store_shared_ex_in_reduction_features]
+      [has_example_header, &shared_example, &ec_seq, &store_shared_ex_in_reduction_features, &data]
       {
         if (has_example_header)
         {
           for (auto& example : ec_seq)
           {
+            if (data.label_type == VW::label_type_t::CB_WITH_OBSERVATIONS) {
```

Comment:
```suggestion
            if (is_cb_with_observations && data.label_type == VW::label_type_t::CB_WITH_OBSERVATIONS) {
```

Reply:
or it might need to be a separate if enclosing both ifs

Reply:
or maybe a separate if enclosing those other ifs because its a compile time boolean